samples: Add apps/pipes.jakt

Lenart12 opened the pull request:
Simple project using inline cpp and system calls to
demonstrate pipe, fork and dup2.

linusg reviewed (changes requested):
https://github.com/SerenityOS/jakt/blob/main/CONTRIBUTING.md#code-submission-policy

> Write the commit message subject line in the imperative mood ("Foo: Change the way dates work", not "Foo: Changed the way dates work").

linusg commented on samples/apps/pipes.jakt:
```diff
@@ -0,0 +1,183 @@
+}
+
+class System {
+    public function getErrno() -> c_int {
```
Jakt follows a similar coding style as serenity, and serenity uses `snake_case` for methods and omits `get_` from getters.

Lenart12 replied:
Function name cant be just `errno` since its defined as a macro. Will rename to `syscall_error_number`

linusg commented on samples/apps/pipes.jakt:
```diff
@@ -0,0 +1,183 @@
+
+class System {
+    public function getErrno() -> c_int {
+        unsafe { cpp {
+            "return errno;"
+        }}
```
```suggestion
        unsafe {
            cpp {
                "return errno;"
            }
        }
```

linusg commented on samples/apps/pipes.jakt:
```diff
@@ -0,0 +1,183 @@
+    }
+
+    public function pipe() throws -> Pipe  {
+        unsafe{
```
Not nearly all of this needs to be wrapped in unsafe. In fact, it looks like *only* the cpp block needs it.

linusg replied:
Also missing a space between the keyword and {

linusg commented on samples/apps/pipes.jakt:
```diff
@@ -0,0 +1,183 @@
+    }
+
+    public function dup2(replace: c_int, with: c_int) throws -> c_int {
+        unsafe{
```
Likewise

linusg commented on samples/apps/pipes.jakt:
```diff
@@ -0,0 +1,183 @@
+    }
+
+    public function fork() throws -> c_int {
+        unsafe{
```
Likewise

linusg commented on samples/apps/pipes.jakt:
```diff
@@ -0,0 +1,183 @@
+                "fork_ret = ::fork();"
+            }
+
+            if fork_ret as! i64 == -1 {
```
Do we have to cast this? `dup2_ret >= 0` works...

Lenart12 replied:
Yes, for now (#465)

linusg commented on samples/apps/pipes.jakt:
```diff
@@ -0,0 +1,183 @@
+    }
+
+    public function wait() throws -> c_int {
+        unsafe{
```
Likewise

linusg replied:
I'll stop pointing these out now, please find the rest yourself :)

linusg commented on samples/apps/pipes.jakt:
```diff
@@ -0,0 +1,183 @@
+
```
Weird newline here

linusg commented on samples/apps/pipes.jakt:
```diff
@@ -0,0 +1,183 @@
+}
+
+function main() {
+
```
Weird newline here

linusg reviewed (changes requested):
Code style comments:

Lenart12 commented:
Changed commit message and implemented changes mentioned in the review. Some weird arithmetic which I don't like was needed because negated c_int needs to be fixed (to avoid things like `not (close_ret >= 0)`, `close_ret == -1` would be better!).

linusg reviewed (changes requested):
`Adds` is not imperative mood

linusg commented on samples/apps/pipes.jakt:
```diff
@@ -0,0 +1,183 @@
+    public function read(fd: c_int, count: usize) throws -> String {
+        let mutable read_ret : c_int = 0;
+
+        unsafe {
```
not all of this needs to be in `unsafe`

Lenart12 commented:
Sorry this is taking so long, english is not my strong suit :(
Unsafe was done like that before because if I had two unsafe blocks the char buffer wouldnt be defined in the second block but I have refactored it so there is only one now.
